File: kotlin/src/test/kotlin/localids/LocalIdTest.kt
package localids

import io.kotest.assertions.fail
import io.kotest.core.spec.style.DescribeSpec
import io.kotest.matchers.comparables.shouldBeLessThan
import io.kotest.matchers.shouldBe
import io.kotest.property.Arb
import io.kotest.property.arbitrary.long
import io.kotest.property.forAll
import kotlin.io.path.Path
import kotlin.time.measureTimedValue

class LocalIdTest : DescribeSpec({

    describe("LocalId.newId()") {
        it("length includes check digit") {
            LocalId.newId().toString().length shouldBe LocalId.CHARACTER_COUNT
        }
        it("with known seed produces known string") {
            LocalId.newId(123456789).toString() shouldBe "4W7gpjVhDs3OMytq"
        }
        it("crude performance test") {
            val time = measureTimedValue {
                repeat(1_000_000) { LocalId.newId() }
            }
            time.duration.inWholeMilliseconds shouldBeLessThan 400
        }
    }

    describe("LocalId.isValid()") {
        it("returns false if the string is the wrong length") {
            LocalId.isValid("0123456789ABCDEFG") shouldBe false
            LocalId.isValid("0123456789ABCDE") shouldBe false
        }
        it("returns false from invalid characters") {
            LocalId.isValid("0123456789ABCDE?") shouldBe false
        }
        it("returns true if the check character is correct") {
            LocalId.isValid("0000000000000000") shouldBe true
            LocalId.isValid("A00000000000000A") shouldBe true
            LocalId.isValid("000b00000000000b") shouldBe true
            LocalId.isValid("000000000X00000X") shouldBe true
            LocalId.isValid("0123456789ABCDEh") shouldBe true
        }
        it("returns false if the check character is incorrect") {
            LocalId.isValid("0123456789ABCDEj") shouldBe false
        }
        it("returns true for many values from a file") {
            val gradleUserDir = System.getProperty("user.dir")
            val ids = gradleUserDir?.let { userDir ->
                Path(userDir, "../valid-ids.txt")
                    .toFile()
                    .readLines()
            }
            if (ids == null) fail("valid-ids.txt not found")
            for (id in ids) {
                LocalId.isValid(id) shouldBe true
            }
        }
    }

    describe("LocalId.hashCode()") {
        it("for two IDs with the same seed are equal") {
            forAll<Long> { seed ->
                LocalId.newId(seed).hashCode() == LocalId.newId(seed).hashCode()
            }
        }
        it("for two IDs with different seeds are not equal") {
            forAll(Arb.long(), Arb.long()) { seed1, seed2 ->
                seed1 == seed2 || LocalId.newId(seed1).hashCode() != LocalId.newId(seed2).hashCode()
            }
        }
    }

    describe("LocalId.equals()") {
        it("for two IDs with the same seed are equal") {
            forAll<Long> { seed ->
                LocalId.newId(seed) == LocalId.newId(seed)
            }
        }
        it("for two IDs with different seeds are not equal") {
            forAll(Arb.long(), Arb.long()) { seed1, seed2 ->
                seed1 == seed2 || LocalId.newId(seed1) != LocalId.newId(seed2)
            }
        }
    }
})

File: javascript/local-id.test.js
const LocalId = require('./local-id');


describe('newId()', () => {
    test('length includes check digit', () => {
        expect(LocalId.newId().length).toEqual(16);
    });
    test('crude performance test', () => {
        let start = performance.now();
        for (let i = 0; i < 100000; i++) {
            LocalId.newId();
        }
        let end = performance.now();
        expect(end - start).toBeLessThan(400);
    });
});

describe('isValid()', () => {
    test('returns false if the string is the wrong length', () => {
        expect(LocalId.isValid("P5O5bH8Hbnm9f4i")).toBe(false);
    });
    test('returns false if the string contains invalid characters', () => {
        expect(LocalId.isValid("@COCqGLPDiCfWeOc")).toBe(false);
    });
    test('returns true if the check character is correct', () => {
        expect(LocalId.isValid("zKy8DqCjTyaiKiSi")).toBe(false);
    });
    test('returns false if the check character is incorrect', () => {
        expect(LocalId.isValid("zKy8DqCjTyaiKiSj")).toBe(false);
    });
});